src/fly/sorts.rs

```diff
@@ -19,100 +20,100 @@ pub enum SortError {
+    UnknownFunctionSort,
+    #[error("could not solve for the sort of {0}")]
+    UnsolvedSort(String),
 }
 
 // entry point for the sort checker
-pub fn check(module: &Module) -> Result<(), (SortError, Option<Span>)> {
-    let build_context = || {
-        let mut sorts = HashSet::new();
-        for sort in &module.signature.sorts {
-            if !sorts.insert(sort.clone()) {
-                return Err(SortError::RedefinedSort(sort.clone()));
-            }
+pub fn check(module: &mut Module) -> Result<(), (SortError, Option<Span>)> {
+    let mut sorts = HashSet::new();
+    for sort in &module.signature.sorts {
+        if !sorts.insert(sort.clone()) {
+            return Err((SortError::RedefinedSort(sort.clone()), None));
         }
+    }
+    let mut vars = UnificationTable::new();
 
-        let mut context = Context {
-            sorts,
-            names: im::HashMap::new(),
-        };
+    let mut context = Context {
+        sorts: &sorts,
+        names: im::HashMap::new(),
+        vars: &mut vars,
+    };
 
+    let mut build_context = || {
         for rel in &module.signature.relations {
             for arg in &rel.args {
-                context.check_sort_exists(arg)?;
+                context.check_sort_exists(arg, false)?;
             }
-            context.check_sort_exists(&rel.sort)?;
+            context.check_sort_exists(&rel.sort, false)?;
             context.add_name(
                 rel.name.clone(),
-                (rel.args.clone(), rel.sort.clone()),
+                AbstractSort::Known(rel.args.clone(), rel.sort.clone()),
                 false,
             )?;
         }
 
-        for def in &module.defs {
+        for def in &mut module.defs {
             {
                 let mut context = context.clone();
-                context.add_binders(&def.binders)?;
-                context.check_sort_exists(&def.ret_sort)?;
-                let ret: Sort = sort_of_term(&context, &def.body)?;
-                sort_eq(&ret, &def.ret_sort)?;
+                context.add_binders(&mut def.binders)?;
+                context.check_sort_exists(&def.ret_sort, false)?;
+                let ret = context.sort_of_term(&mut def.body)?;
+                context.sort_eq(&ret, &AbstractSort::unit(def.ret_sort.clone()))?;
             }
 
             let args = def
                 .binders
                 .iter()
                 .map(|binder| binder.sort.clone())
                 .collect();
-            context.add_name(def.name.clone(), (args, def.ret_sort.clone()), false)?;
+
+            context.add_name(
+                def.name.clone(),
+                AbstractSort::Known(args, def.ret_sort.clone()),
+                false,
+            )?;
         }
 
-        Ok(context)
+        Ok(())
     };
 
-    let context = match build_context() {
+    match build_context() {
         Ok(context) => context,
         Err(e) => return Err((e, None)),
     };
```

Comment:
I'm confused about this change - isn't its value just ignored now? The `Ok(context) => context` doesn't appear to do anything without the `let context` binding.

Separately, you can probably use `Result::wrap_err()` combined with `?` to make this a bit cleaner.

Reply:
Yep, it doesn't do anything. It's left over from when build_context returned the context. I don't really understand wrap_err (from eyre, not std, right?) so I went for just using map_err instead.

Reply:
I actually meant `map_err`, glad you figured that out! eyre is mostly intended for easy application errors that have good debugging context. For libraries it's recommended to use more structured error messages like you're currently doing.